src/lib/child_processes/child_processes.ml

```diff
@@ -0,0 +1,444 @@
+ fun t -> Ivar.peek t.terminated_ivar
+
+(** Try running [f] until it returns [Ok], returning the first [Ok] or [Error]
+    if all attempts fail. *)
+let keep_trying :
+    f:('a -> 'b Deferred.Or_error.t) -> 'a list -> 'b Deferred.Or_error.t =
+ fun ~f xs ->
+  let open Deferred.Let_syntax in
+  let rec go e xs : 'b Deferred.Or_error.t =
+    match xs with
+    | [] ->
+        return e
+    | x :: xs -> (
+        match%bind f x with
+        | Ok r ->
+            return (Ok r)
+        | Error e ->
+            go (Error e) xs )
+  in
+  go (Or_error.error_string "empty input") xs
+
+(* Unfortunately, `dune runtest` runs in a pwd deep inside the build directory.
+   This hack finds the project root by recursively looking for the dune-project
+   file. *)
+let get_project_root () =
+  let open Filename in
+  let rec go dir =
+    if Core.Sys.file_exists_exn @@ dir ^/ "src/dune-project" then Some dir
+    else if String.equal dir "/" then None
+    else go @@ fst @@ split dir
+  in
+  go @@ realpath current_dir_name
+
+(* This snippet was taken from our fork of RPC Parallel. Would be nice to have a
+   shared utility, but this is easiest for now. *)
+(* To get the currently running executable:
+   On Darwin:
+   Use _NSGetExecutablePath via Ctypes
+
+   On Linux:
+   Use /proc/PID/exe
+   - argv[0] might have been deleted (this is quite common with jenga)
+   - `cp /proc/PID/exe dst` works as expected while `cp /proc/self/exe dst` does
+     not *)
+let get_coda_binary () =
+  let open Async in
+  let open Deferred.Or_error.Let_syntax in
+  let%bind os = Process.run ~prog:"uname" ~args:["-s"] () in
+  if os = "Darwin\n" then
+    let open Ctypes in
+    let ns_get_executable_path =
+      Foreign.foreign "_NSGetExecutablePath"
+        (ptr char @-> ptr uint32_t @-> returning int)
+    in
+    let path_max = Syslimits.path_max () in
+    let buf = Ctypes.allocate_n char ~count:path_max in
+    let count = Ctypes.allocate uint32_t (Unsigned.UInt32.of_int path_max) in
+    let%map () =
+      Deferred.return
+        (Result.ok_if_true
+           (ns_get_executable_path buf count = 0)
+           ~error:
+             (Error.of_string
+                "call to _NSGetExecutablePath failed unexpectedly"))
+    in
+    let s = string_from_ptr buf ~length:(!@count |> Unsigned.UInt32.to_int) in
+    List.hd_exn @@ String.split s ~on:(Char.of_int 0 |> Option.value_exn)
+  else
+    (* FIXME for finding the executable relative to the install path this should
+       deference the symlink if possible. *)
+    Deferred.Or_error.return
+      (Unix.getpid () |> Pid.to_int |> sprintf "/proc/%d/exe")
+
+(* Check the PID file, and if it exists and corresponds to a currently running
+   process, kill that process. This runs when the daemon starts, and should
+   *not* be used to kill a process that was started during this run of the
+   daemon.
+*)
+let maybe_kill_and_unlock : string -> string -> Logger.t -> unit Deferred.t =
```

Comment:
can `lockpath` be of type `Core_kernel.Filename.t` (which is just a string, but makes the role of the argument clear)

Reply:
:heavy_check_mark: